### PR TITLE
Fix bug where enums with underlying types would cause a crash

### PR DESCRIPTION
**Pull request by Genbox**
I've added a unit test to demonstrate the issue. The pattern-based check in EnumWithoutDefaultValueAnalyzer.cs:49 only works when the enum is using its default underlying type (int). When it is any other numerical type (byte, short, ushort, uint, long or ulong) it crashes due to an invalid cast.

`Convert.ToULong()` works because it uses the IConvertible interface on the numerical types to convert between values, and using `ToUlong()` instead of f.ex. `ToInt()` means we don't have to check for all possible types, since we are only interested in if the value is 0.

### CODE REVIEW

**Comment by Genbox** [comment]
I just tested with negative values and it does not work.

File: SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs [hunk 1]
```diff
@@ -81,6 +81,33 @@ public async Task EnumWithoutDefaultValue_RightName_RightValueDuplicated(string
...
+        await VerifyCS.VerifyNoDiagnostic(original);
+    }
+
+    [TestMethod]
+    [DataRow("None")]
+    [DataRow("Unknown")]
+    public async Task EnumWithoutDefaultValue_RightName_NegativeValue(string memberName)
+    {
+        var original = $@"
+enum [|Test|] : short {{
```
Comment (Vannevelj) [inline SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs:95]:
```suggestion
enum Test : short {{
```

Those are used to indicate where the diagnostic is but considering we don't have any, we can omit it

File: SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs [hunk 2]
```diff
@@ -81,6 +81,33 @@ public async Task EnumWithoutDefaultValue_RightName_RightValueDuplicated(string
...
+    {memberName} = -1
+}}";
+
+        await VerifyCS.VerifyNoDiagnostic(original);
+    }
+
+    [TestMethod]
+    [DataRow("None")]
+    [DataRow("Unknown")]
+    public async Task EnumWithoutDefaultValue_RightName_WithDifferentType(string memberName)
```
Comment (Vannevelj) [inline SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs:105]: Can you make this a parameterised test that has a `[DataRow]` for each type an enum can take? That way this unit test is exhaustive of the different combinations. I believe that means `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, and `ulong`.

I think you can use a constant `memberName` here to simplify the test

**Review by Vannevelj — COMMENTED** [review]
Good find! I've left a few comments to address in the unit tests and solution looks good to me otherwise.

File: SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs (hunk 2 above)
Comment (Genbox) [reply SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs]: Sorry for the delay. Had to get home from work first.

Only just now saw your comment about simplifying it. I'll let it be up to you if you still want it. :)

File: SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs (hunk 1 above)
Comment (Genbox) [reply SharpSource/SharpSource.Test/EnumWithoutDefaultValueTests.cs:96]: There is a diagnostic. The value is not 0 as it should be.

The test ensures that the conversion code in the analyzer don't fail on negative values, but do indeed report as it should.

**Pull request merged**